src/log-pcap.c

```diff
@@ -463,6 +497,85 @@ static void PcapLogUnlock(PcapLogData *pl)
     }
 }
 
+static inline int PcapWrite(
+        PcapLogData *pl, PcapLogCompressionData *comp, uint8_t *data, size_t len)
+{
+    pcap_dump((u_char *)pl->pcap_dumper, pl->h, data);
+    if (pl->compression.format == PCAP_LOG_COMPRESSION_FORMAT_NONE) {
+        pl->size_current += len;
+    }
+#ifdef HAVE_LIBLZ4
+    else if (pl->compression.format == PCAP_LOG_COMPRESSION_FORMAT_LZ4) {
+        pcap_dump_flush(pl->pcap_dumper);
+        uint64_t in_size = (uint64_t)ftell(comp->pcap_buf_wrapper);
+        uint64_t out_size = LZ4F_compressUpdate(
+                comp->lz4f_context, comp->buffer, comp->buffer_size, comp->pcap_buf, in_size, NULL);
+        if (LZ4F_isError(len)) {
+            SCLogError(SC_ERR_PCAP_LOG_COMPRESS, "LZ4F_compressUpdate: %s", LZ4F_getErrorName(len));
+            return TM_ECODE_FAILED;
+        }
+        if (fseek(pl->compression.pcap_buf_wrapper, 0, SEEK_SET) != 0) {
+            SCLogError(SC_ERR_FSEEK, "fseek failed: %s", strerror(errno));
+            return TM_ECODE_FAILED;
+        }
+        if (fwrite(comp->buffer, 1, out_size, comp->file) < out_size) {
+            SCLogError(SC_ERR_FWRITE, "fwrite failed: %s", strerror(errno));
+            return TM_ECODE_FAILED;
+        }
+        if (out_size > 0) {
+            pl->size_current += out_size;
+            comp->bytes_in_block = len;
+        } else {
+            comp->bytes_in_block += len;
+        }
+    }
+#endif /* HAVE_LIBLZ4 */
+    return TM_ECODE_OK;
+}
+
+struct PcapLogCallbackContext {
+    PcapLogData *pl;
+    PcapLogCompressionData *connp;
+    MemBuffer *buf;
+};
+
+static int PcapLogSegmentCallback(
+        const Packet *p, TcpSegment *seg, void *data, const uint8_t *buf, uint32_t buflen)
+{
+    struct PcapLogCallbackContext *pctx = (struct PcapLogCallbackContext *)data;
+
+    if (seg->pcap_hdr_storage->pktlen) {
+        pctx->pl->h->ts.tv_sec = seg->pcap_hdr_storage->ts.tv_sec;
+        pctx->pl->h->ts.tv_usec = seg->pcap_hdr_storage->ts.tv_usec;
+        pctx->pl->h->len = seg->pcap_hdr_storage->pktlen + buflen;
+        pctx->pl->h->caplen = seg->pcap_hdr_storage->pktlen + buflen;
+        MemBufferReset(pctx->buf);
+        MemBufferWriteRaw(pctx->buf, seg->pcap_hdr_storage->pkt_hdr, seg->pcap_hdr_storage->pktlen);
+        MemBufferWriteRaw(pctx->buf, buf, buflen);
+
+        PcapWrite(pctx->pl, pctx->connp, (uint8_t *)pctx->buf->buffer,
+                seg->pcap_hdr_storage->pktlen + buflen);
+    }
+    return 1;
+}
+
+static void PcapLogDumpSegments(
```

Comment:
Still working on a full review, but playing around with your branch a bit... It looks like this is only used in a single direction correct? Comparing outputs between our two approaches it looks like this results in a few missing packets. See https://github.com/scottfgjordan/suricata/blob/3af87b26638cfcc7b64eee532a54bbebbc32dc62/src/detect-tag-pcap.c#L244 for how I went about this. I think otherwise the results between the two should be pretty similar.

Reply:
Yes, this patchset just output one side that will contain the data that did trigger the alert. I still need to be convinced we can have useful data when dumping both sides.

Reply:
I did a small test just comparing the capture results for a basic signature that would alert on an HTTP GET request to compare how the results differ between the two approaches. Links to the captures can be found here: https://wiki.wireshark.org/SampleCaptures (tcp-ecn-sample.pcap and http.pcap). The signature I used was simple: alert tcp any any -> any any (msg:"ALERTING ON HTTP GET REQUEST"; http.method; content:"GET"; tag:session; sid:1; rev:1;).

The ideal results in the situation would be a capture starting from the packet with the GET request and then capturing all packets in the session from that point. The results are as follows:

For Pcap conditional v2.0.0
- Conditional: alerts
  - tcp-ecn-sample.pcap: Grabs the packet that caused the alert (Note: there is a duplication of the packet that causes the alert) then misses the next two packets immediately after this (in the opposite direction) which contain the HTTP OK response. 
  - http.pcap: Grabs the packet that caused the alert (duplicated again) and then misses the next 3 packets (2 in the opposite direction of alert and 1 in the same) resulting in losing out on the HTTP OK response and some additional data.
- Conditional: tag
  - tcp-ecn-sample.pcap: Misses the packet that causes the alert and the following 2 packets. Picks it up from there.
  - http.pcap: Misses the packet that causes the alert and the following 3 packets. Picks it up from there. 
For capture feature 120 v5
No Session Dumping:
- tcp-ecn-sample.pcap:  Misses the packet that causes the alert and the following packet. Picks it up from there
- http.pcap: Misses the packet that causes the alert and the following 2 packets. Picks it up from there. 
Session Dumping Enabled:
- tcp-ecn-sample.pcap: Grabs the packet that causes the alert and everything following.
- http.pcap: Grabs the packet that causes the alert and everything following. 

A small scale example, but hopefully does a bit of convincing you of some the value with dumping the tcp session and the value of doing both directions.

Reply:
Ok, thanks a lot for feedback and tests. I'm gonna replay them and try to see what is needed.